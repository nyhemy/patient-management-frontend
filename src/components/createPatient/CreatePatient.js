import React, { useState } from 'react';
import { useHistory } from 'react-router-dom';
import loadImg from '../ajax-loader.gif';
import styles from './CreatePatient.module.css';
import { emailRegex, zipcodeRegex, ssnRegex } from '../Constants';
import { stateValidator, genderValidator } from '../Functions';

const axios = require('axios').default;

const CreatePatient = () => {
  // states used for general component functionality
  const history = useHistory();
  const [loading, setLoading] = useState(false);
  const [errorMsg, setErrorMsg] = useState('');

  // states used for form inputs
  const [firstName, setFirstName] = useState('');
  const [lastName, setLastName] = useState('');
  const [ssn, setSsn] = useState('');
  const [email, setEmail] = useState('');
  const [street, setStreet] = useState('');
  const [city, setCity] = useState('');
  const [state, setState] = useState('');
  const [zipcode, setZipCode] = useState('');
  const [age, setAge] = useState(0);
  const [height, setHeight] = useState(0);
  const [weight, setWeight] = useState('');
  const [insurance, setInsurance] = useState('');
  const [gender, setGender] = useState('');

  // states used for error handling
  const [firstNameError, setFirstNameError] = useState('');
  const [lastNameError, setLastNameError] = useState('');
  const [ssnError, setSsnError] = useState('');
  const [emailError, setEmailError] = useState('');
  const [streetError, setStreetError] = useState('');
  const [cityError, setCityError] = useState('');
  const [stateError, setStateError] = useState('');
  const [zipcodeError, setZipCodeError] = useState('');
  const [ageError, setAgeError] = useState('');
  const [heightError, setHeightError] = useState('');
  const [weightError, setWeightError] = useState('');
  const [insuranceError, setInsuranceError] = useState('');
  const [genderError, setGenderError] = useState('');

  /**
   * clears all error states
   */
  const clearErrors = () => {
    setFirstNameError('');
    setLastNameError('');
    setSsnError('');
    setEmailError('');
    setStreetError('');
    setCityError('');
    setStateError('');
    setZipCodeError('');
    setAgeError('');
    setHeightError('');
    setWeightError('');
    setInsuranceError('');
    setGenderError('');
  };

  /**
   * Handles changes to inputs
   *
   * @param {event} event is when input changes its value
   */
  const handleChange = (event) => {
    switch (event.target.name) {
      case 'firstName':
        setFirstName(event.target.value);
        break;
      case 'lastName':
        setLastName(event.target.value);
        break;
      case 'ssn':
        setSsn(event.target.value);
        break;
      case 'email':
        setEmail(event.target.value);
        break;
      case 'street':
        setStreet(event.target.value);
        break;
      case 'city':
        setCity(event.target.value);
        break;
      case 'state':
        setState(event.target.value);
        break;
      case 'zipcode':
        setZipCode(event.target.value);
        break;
      case 'age':
        setAge(event.target.value);
        break;
      case 'height':
        setHeight(event.target.value);
        break;
      case 'weight':
        setWeight(event.target.value);
        break;
      case 'insurance':
        setInsurance(event.target.value);
        break;
      case 'gender':
        setGender(event.target.value);
        break;
      default:
        break;
    }
  };

  /**
   * Handles form submission event, including validaion and API calls
   *
   * @param {event} event is the form submission event
   */
  const handleSubmit = (event) => {
    event.preventDefault();
    clearErrors();
    let noValidate = false;

    if (firstName === '' || !firstName.trim().length) {
      setFirstNameError('Must be valid first name');
      noValidate = true;
    }

    if (lastName === '' || !lastName.trim().length) {
      setLastNameError('Must be valid last name');
      noValidate = true;
    }

    if (ssn === '' || !ssn.trim().length || !ssnRegex.test(ssn)) {
      setSsnError('Must be valid ssn');
      noValidate = true;
    }

    if (email === '' || !email.trim().length || !emailRegex.test(email)) {
      setEmailError('Must be valid email');
      noValidate = true;
    }

    if (street === '' || !street.trim().length) {
      setStreetError('Must be valid street');
      noValidate = true;
    }

    if (city === '' || !city.trim().length) {
      setCityError('Must be valid city');
      noValidate = true;
    }

    if (state === '' || !state.trim().length || !stateValidator(state)) {
      setStateError('Must be valid state');
      noValidate = true;
    }

    if (zipcode === '' || !zipcode.trim().length || !zipcodeRegex.test(zipcode)) {
      setZipCodeError('Must be valid zipcode');
      noValidate = true;
    }

    if (age <= 0 || Number.isNaN(Number(age))) {
      setAgeError('Must be valid age');
      noValidate = true;
    }

    if (height <= 0 || Number.isNaN(Number(height))) {
      setHeightError('Must be valid height');
      noValidate = true;
    }

    if (weight <= 0 || Number.isNaN(Number(weight))) {
      setWeightError('Must be valid weight');
      noValidate = true;
    }

    if (insurance === '' || !insurance.trim().length) {
      setInsuranceError('Must be valid insurance');
      noValidate = true;
    }

    if (gender === '' || !gender.trim().length || !genderValidator(gender)) {
      setGenderError('Must be valid gender');
      noValidate = true;
    }

    if (noValidate) {
      return;
    }

    setLoading(true);

    axios.post('http://localhost:8080/patients',
      {
        firstName,
        lastName,
        ssn,
        email,
        street,
        city,
        state,
        postal: zipcode,
        age,
        height,
        weight,
        insurance,
        gender
      },
      {
        headers: {
          'Content-Type': 'application/json',
          mode: 'cors',
          Authorization: `Bearer ${sessionStorage.getItem('token')}`
        }
      })
      // eslint-disable-next-line no-unused-vars
      .then((response) => {
        setLoading(false);
        history.push('/patients');
      })
      // eslint-disable-next-line no-unused-vars
      .catch((error) => {
        setLoading(false);
        setErrorMsg('Oops something went wrong');
      });
  };

  return (
    <div className={styles.container}>
      <h2>Create Patient</h2>
      <h3 className={styles.error}>
        {loading
          ? <img src={loadImg} alt="loading..." />
          : errorMsg}
      </h3>
      <form className={styles.form} onSubmit={handleSubmit} noValidate>
        <div className={styles.input}>
          <input type="text" name="firstName" placeholder="first name" onChange={handleChange} />
          {' '}
          <input type="text" name="lastName" placeholder="last name" onChange={handleChange} />
        </div>
        <div className={styles.inputError}>
          <div className={styles.inputErrorLeft}>{firstNameError}</div>
          {' '}
          <div className={styles.inputErrorRight}>{lastNameError}</div>
        </div>

        <div className={styles.input}>
          <input type="text" name="ssn" placeholder="social security number" onChange={handleChange} />
          {' '}
          <input type="email" name="email" placeholder="email" onChange={handleChange} />
        </div>
        <div className={styles.inputError}>
          <span className={styles.inputErrorLeft}>{ssnError}</span>
          {' '}
          <span className={styles.inputErrorRight}>{emailError}</span>
        </div>

        <div className={styles.input}>
          <input type="text" name="street" placeholder="street" onChange={handleChange} />
          {' '}
          <input type="text" name="city" placeholder="city" onChange={handleChange} />
        </div>
        <div className={styles.inputError}>
          <span className={styles.inputErrorLeft}>{streetError}</span>
          {' '}
          <span className={styles.inputErrorRight}>{cityError}</span>
        </div>

        <div className={styles.input}>
          <input type="text" name="state" placeholder="state" onChange={handleChange} />
          {' '}
          <input type="text" name="zipcode" placeholder="zipcode" onChange={handleChange} />
        </div>
        <div className={styles.inputError}>
          <span className={styles.inputErrorLeft}>{stateError}</span>
          {' '}
          <span className={styles.inputErrorRight}>{zipcodeError}</span>
        </div>

        <div className={styles.input}>
          <input type="number" name="age" placeholder="age" onChange={handleChange} />
          {' '}
          <input type="number" name="height" placeholder="height" onChange={handleChange} />
        </div>
        <div className={styles.inputError}>
          <span className={styles.inputErrorLeft}>{ageError}</span>
          {' '}
          <span className={styles.inputErrorRight}>{heightError}</span>
        </div>

        <div className={styles.input}>
          <input type="number" name="weight" placeholder="weight" onChange={handleChange} />
          {' '}
          <input type="text" name="insurance" placeholder="insurance" onChange={handleChange} />
        </div>
        <div className={styles.inputError}>
          <span className={styles.inputErrorLeft}>{weightError}</span>
          {' '}
          <span className={styles.inputErrorRight}>{insuranceError}</span>
        </div>

        <div>
          {/* eslint-disable-next-line jsx-a11y/no-onchange */}
          <select defaultValue="DEFAULT" className={styles.select} name="gender" onChange={handleChange}>
            <option value="DEFAULT">--select gender--</option>
            <option value="male">Male</option>
            <option value="memale">Female</option>
            <option value="other">Other</option>
          </select>
        </div>
        <div className={styles.inputError}>{genderError}</div>

        <button type="submit">Create</button>
      </form>
    </div>
  );
};

export default CreatePatient;
