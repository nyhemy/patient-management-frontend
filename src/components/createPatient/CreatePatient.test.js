import React from 'react';
import { cleanup, render, screen, act, waitFor, getByText, queryByText, fireEvent, getByAltText, getByTestId, queryByTestId } from '@testing-library/react';
import { createMemoryHistory } from 'history';
import { Router, Route, MemoryRouter } from 'react-router-dom';
import renderer from 'react-test-renderer';
import fetchMock from 'jest-fetch-mock';
import userEvent from '@testing-library/user-event';
import CreatePatient from './CreatePatient';
import '@testing-library/jest-dom/extend-expect';

// const history = createMemoryHistory();

fetchMock.enableMocks();

beforeEach(() => {
  fetch.resetMocks();
});

afterEach(cleanup);

const mockHistoryPush = jest.fn();

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useHistory: () => ({
    push: mockHistoryPush,
  }),
}));

const fillInputData = () => {
  userEvent.type(screen.getByTestId('f-name'), 'test');
  userEvent.type(screen.getByTestId('l-name'), 'test');
  userEvent.type(screen.getByTestId('ssn'), '000-00-0000');
  userEvent.type(screen.getByTestId('email'), 'test@test.test');
  userEvent.type(screen.getByTestId('street'), 'test');
  userEvent.type(screen.getByTestId('city'), 'test');
  userEvent.type(screen.getByTestId('state'), 'MA');
  userEvent.type(screen.getByTestId('zipcode'), '00000');
  userEvent.type(screen.getByTestId('age'), '1');
  userEvent.type(screen.getByTestId('height'), '1');
  userEvent.type(screen.getByTestId('weight'), '1');
  userEvent.type(screen.getByTestId('insurance'), 'Test');
  userEvent.selectOptions(screen.getByTestId('gender-select'), ['male']);
};

// const patientDataMock = JSON.stringify(
//   {
//     id: 1,
//     firstName: "Test",
//     lastName: "Last",
//     email: "tl@gmail.com",
//     street: "Yo Dr",
//     city: "Heyvalley",
//     state: "NH",
//     postal: "00000",
//     age: 23,
//     height: 64,
//     weight: 112,
//     insurance: "Wooshoo Inc",
//     gender: "female",
//     ssn: "000-00-0000"
//   }
// );

it('renders and matches snapshot', () => {
  render(<CreatePatient />);
  expect(screen.queryByText('Create Patient')).toBeTruthy();

  const component = renderer.create(<CreatePatient />);
  const tree = component.toJSON();

  expect(tree).toMatchSnapshot();
});

it('updates patient data', async () => {
  fetch
    .mockResponseOnce({ status: 200, method: 'POST', headers: { 'content-type': 'application/json' } });

  const { getByText, queryByText, getByTestId, getByAltText } = render(<CreatePatient />);

  fillInputData();

  const selectButton = getByText('Create');
  fireEvent.click(selectButton);
  await waitFor(() => getByAltText('loading...'));
  expect(mockHistoryPush).toHaveBeenCalledWith('/patients');
});

it('fails post request', async () => {
  fetch
    .mockResponse('{ "id": 1 }', { status: 409, headers: { 'content-type': 'application/json' } });

  const { getByText, queryByText, getByTestId, getByAltText } = render(<CreatePatient />);

  fillInputData();

  const selectButton = getByText('Create');
  fireEvent.click(selectButton);
  await waitFor(() => getByAltText('loading...'));
  expect(queryByText('Conflicts with existing data, likely email')).toBeTruthy();
});

it('catches generic error', async () => {
  fetch
    .mockReject(new Error('error'));

  const { getByText, queryByText, getByTestId, getByAltText } = render(<CreatePatient />);

  fillInputData();

  const selectButton = getByText('Create');
  fireEvent.click(selectButton);
  await waitFor(() => getByAltText('loading...'));
  expect(queryByText('Oops something went wrong')).toBeTruthy();
});

it('tests if input error messages appear', async () => {
  const { getByText, queryByText } = render(<CreatePatient />);

  const selectButton = getByText('Create');
  fireEvent.click(selectButton);
  await waitFor(() => getByText('Must be valid first name'));
  expect(queryByText('Must be valid first name')).toBeTruthy();
  expect(queryByText('Must be valid last name')).toBeTruthy();
  expect(queryByText('Must be valid ssn')).toBeTruthy();
  expect(queryByText('Must be valid email')).toBeTruthy();
  expect(queryByText('Must be valid street')).toBeTruthy();
  expect(queryByText('Must be valid city')).toBeTruthy();
  expect(queryByText('Must be valid state')).toBeTruthy();
  expect(queryByText('Must be valid zipcode')).toBeTruthy();
  expect(queryByText('Must be valid age')).toBeTruthy();
  expect(queryByText('Must be valid height')).toBeTruthy();
  expect(queryByText('Must be valid weight')).toBeTruthy();
  expect(queryByText('Must be valid insurance')).toBeTruthy();
  expect(queryByText('Must be valid gender')).toBeTruthy();
});
