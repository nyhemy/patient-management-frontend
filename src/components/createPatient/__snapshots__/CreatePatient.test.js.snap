// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders and matches snapshot 1`] = `
<div
  className="container"
>
  <h2>
    Create Patient
  </h2>
  <h3
    className="error"
  >
    
  </h3>
  <form
    className="form"
    noValidate={true}
    onSubmit={[Function]}
  >
    <div
      className="input"
    >
      <input
        data-testid="f-name"
        name="firstName"
        onChange={[Function]}
        placeholder="first name"
        type="text"
        value=""
      />
       
      <input
        data-testid="l-name"
        name="lastName"
        onChange={[Function]}
        placeholder="last name"
        type="text"
        value=""
      />
    </div>
    <div
      className="inputError"
    >
      <div
        className="inputErrorLeft"
      >
        
      </div>
       
      <div
        className="inputErrorRight"
      >
        
      </div>
    </div>
    <div
      className="input"
    >
      <input
        data-testid="ssn"
        name="ssn"
        onChange={[Function]}
        placeholder="social security number"
        type="text"
        value=""
      />
       
      <input
        data-testid="email"
        name="email"
        onChange={[Function]}
        placeholder="email"
        type="email"
        value=""
      />
    </div>
    <div
      className="inputError"
    >
      <span
        className="inputErrorLeft"
      >
        
      </span>
       
      <span
        className="inputErrorRight"
      >
        
      </span>
    </div>
    <div
      className="input"
    >
      <input
        data-testid="street"
        name="street"
        onChange={[Function]}
        placeholder="street"
        type="text"
        value=""
      />
       
      <input
        data-testid="city"
        name="city"
        onChange={[Function]}
        placeholder="city"
        type="text"
        value=""
      />
    </div>
    <div
      className="inputError"
    >
      <span
        className="inputErrorLeft"
      >
        
      </span>
       
      <span
        className="inputErrorRight"
      >
        
      </span>
    </div>
    <div
      className="input"
    >
      <input
        data-testid="state"
        name="state"
        onChange={[Function]}
        placeholder="state"
        type="text"
        value=""
      />
       
      <input
        data-testid="zipcode"
        name="zipcode"
        onChange={[Function]}
        placeholder="zipcode"
        type="text"
        value=""
      />
    </div>
    <div
      className="inputError"
    >
      <span
        className="inputErrorLeft"
      >
        
      </span>
       
      <span
        className="inputErrorRight"
      >
        
      </span>
    </div>
    <div
      className="input"
    >
      <input
        data-testid="age"
        name="age"
        onChange={[Function]}
        placeholder="age"
        type="number"
        value={0}
      />
       
      <input
        data-testid="height"
        name="height"
        onChange={[Function]}
        placeholder="height"
        type="number"
        value={0}
      />
    </div>
    <div
      className="inputError"
    >
      <span
        className="inputErrorLeft"
      >
        
      </span>
       
      <span
        className="inputErrorRight"
      >
        
      </span>
    </div>
    <div
      className="input"
    >
      <input
        data-testid="weight"
        name="weight"
        onChange={[Function]}
        placeholder="weight"
        type="number"
        value=""
      />
       
      <input
        data-testid="insurance"
        name="insurance"
        onChange={[Function]}
        placeholder="insurance"
        type="text"
        value=""
      />
    </div>
    <div
      className="inputError"
    >
      <span
        className="inputErrorLeft"
      >
        
      </span>
       
      <span
        className="inputErrorRight"
      >
        
      </span>
    </div>
    <div>
      <select
        className="select"
        data-testid="gender-select"
        defaultValue="DEFAULT"
        name="gender"
        onChange={[Function]}
        value=""
      >
        <option
          data-testid="default-select"
          value="DEFAULT"
        >
          --select gender--
        </option>
        <option
          data-testid="male-select"
          value="male"
        >
          Male
        </option>
        <option
          data-testid="female-select"
          value="female"
        >
          Female
        </option>
        <option
          data-testid="other-select"
          value="other"
        >
          Other
        </option>
      </select>
    </div>
    <div
      className="inputError"
    >
      
    </div>
    <button
      type="submit"
    >
      Create
    </button>
  </form>
</div>
`;
